feat: adjust trace timestamps with internal clock

vigneshshanmugam opened the pull request:
+ Part of #85 
+ converts the trace timestamp used by the chrome devtools trace event format to the internally used monotonic clock which is based on the network events and also used on journey and step start/end timings.
+ This makes querying for filmstrips for each step easier as we can directly query for all `journey/filmstrip` events within the `start` and `end` time of the step.

apmmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-157/detail/PR-157/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-157/detail/PR-157/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-157/detail/PR-157/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-157/detail/PR-157/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://synthetics_157.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `Pull request #157 opened`

* Start Time: 2020-12-02T11:41:18.644+0000

* Duration: 6 min 14 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 43  |
| Skipped      | 0 |
| Total        | 43   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-157/detail/PR-157/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>

  
  
  
  
  
#### `Runs the tests`
<ul>
<li>Took 0 min 19 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/elastic-synthetics/pipelines/PR-157/runs/1/steps/210/log/?start=0">here</a></li>
<li>Description: <code>./node_modules/.bin/jest --ci --reporters=default --reporters=jest-junit</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-12-02T11:46:39.830Z] [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
[2020-12-02T11:46:39.830Z] 
[2020-12-02T11:46:41.765Z] => Close and reopen your terminal to start using nvm or run the following to use it now:
[2020-12-02T11:46:41.765Z] 
[2020-12-02T11:46:41.765Z] export NVM_DIR="$HOME/.nvm"
[2020-12-02T11:46:41.765Z] [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
[2020-12-02T11:46:41.765Z] Installing latest LTS version.
[2020-12-02T11:46:41.765Z] v14.15.1 is already installed.
[2020-12-02T11:46:42.040Z] Now using node v14.15.1 (npm v6.14.8)
[2020-12-02T11:46:42.188Z] Running in /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-157/src/github.com/elastic/synthetics
[2020-12-02T11:46:42.546Z] + npm run build
[2020-12-02T11:46:42.546Z] 
[2020-12-02T11:46:42.546Z] > @elastic/synthetics@0.0.1-alpha.6 build /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-157/src/github.com/elastic/synthetics
[2020-12-02T11:46:42.546Z] > tsc
[2020-12-02T11:46:42.546Z] 
[2020-12-02T11:46:47.046Z] Running in /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-157/src/github.com/elastic/synthetics
[2020-12-02T11:47:01.309Z] + set +x
[2020-12-02T11:47:01.309Z] => nvm is already installed in /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-157/.nvm, trying to update using git
[2020-12-02T11:47:01.876Z] 
=> => Compressing and cleaning up git repository
[2020-12-02T11:47:01.876Z] 
[2020-12-02T11:47:01.876Z] => Profile not found. Tried ~/.bashrc, ~/.bash_profile, ~/.zshrc, and ~/.profile.
[2020-12-02T11:47:01.876Z] => Create one of them and run this script again
[2020-12-02T11:47:01.876Z]    OR
[2020-12-02T11:47:01.876Z] => Append the following lines to the correct file yourself:
[2020-12-02T11:47:01.876Z] 
[2020-12-02T11:47:01.876Z] export NVM_DIR="$HOME/.nvm"
[2020-12-02T11:47:01.876Z] [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
[2020-12-02T11:47:01.876Z] 
[2020-12-02T11:47:03.779Z] => Close and reopen your terminal to start using nvm or run the following to use it now:
[2020-12-02T11:47:03.779Z] 
[2020-12-02T11:47:03.779Z] export NVM_DIR="$HOME/.nvm"
[2020-12-02T11:47:03.779Z] [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
[2020-12-02T11:47:03.779Z] Installing latest LTS version.
[2020-12-02T11:47:03.779Z] v14.15.1 is already installed.
[2020-12-02T11:47:04.038Z] Now using node v14.15.1 (npm v6.14.8)
[2020-12-02T11:47:04.129Z] Running in /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-157/src/github.com/elastic/synthetics
[2020-12-02T11:47:04.444Z] + npm add --dev jest-junit
[2020-12-02T11:47:04.703Z] npm WARN install Usage of the `--dev` option is deprecated. Use `--also=dev` instead.
[2020-12-02T11:47:09.971Z] npm WARN @elastic/synthetics@0.0.1-alpha.6 No repository field.
[2020-12-02T11:47:09.971Z] npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.1.3 (node_modules/fsevents):
[2020-12-02T11:47:09.971Z] npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
[2020-12-02T11:47:09.971Z] 
[2020-12-02T11:47:09.971Z] + jest-junit@12.0.0
[2020-12-02T11:47:09.971Z] added 4 packages from 13 contributors and audited 759 packages in 5.319s
[2020-12-02T11:47:10.538Z] 
[2020-12-02T11:47:10.538Z] 44 packages are looking for funding
[2020-12-02T11:47:10.538Z]   run `npm fund` for details
[2020-12-02T11:47:10.538Z] 
[2020-12-02T11:47:10.538Z] found 0 vulnerabilities
[2020-12-02T11:47:10.538Z] 
[2020-12-02T11:47:10.837Z] + ./node_modules/.bin/jest --ci --reporters=default --reporters=jest-junit
[2020-12-02T11:47:18.978Z] FAIL __tests__/reporters/json.test.ts
[2020-12-02T11:47:18.978Z]   ● Test suite failed to run
[2020-12-02T11:47:18.978Z] 
[2020-12-02T11:47:18.978Z]     __tests__/reporters/json.test.ts:110:11 - error TS2322: Type '{ snapshot: string; name: string; ts: number; }' is not assignable to type 'FilmStrip'.
[2020-12-02T11:47:18.978Z]       Object literal may only specify known properties, and 'name' does not exist in type 'FilmStrip'.
[2020-12-02T11:47:18.978Z] 
[2020-12-02T11:47:18.978Z]     110           name: 'screenshot',
[2020-12-02T11:47:18.978Z]                   ~~~~~~~~~~~~~~~~~~
[2020-12-02T11:47:18.978Z] 
[2020-12-02T11:47:18.978Z] PASS __tests__/helper.test.ts (7.142 s)
[2020-12-02T11:47:20.885Z] PASS __tests__/core/index.test.ts
[2020-12-02T11:47:21.831Z] PASS __tests__/index.test.ts
[2020-12-02T11:47:21.831Z] PASS __tests__/reporters/base.test.ts
[2020-12-02T11:47:23.742Z] PASS __tests__/plugins/browser-console.test.ts
[2020-12-02T11:47:24.000Z] PASS __tests__/plugins/performance.test.ts
[2020-12-02T11:47:24.937Z] PASS __tests__/plugins/plugin-manager.test.ts
[2020-12-02T11:47:26.849Z] PASS __tests__/plugins/tracing.test.ts
[2020-12-02T11:47:27.789Z] PASS __tests__/core/gatherer.test.ts
[2020-12-02T11:47:28.736Z] PASS __tests__/plugins/network.test.ts
[2020-12-02T11:47:28.994Z] PASS __tests__/core/logger.test.ts
[2020-12-02T11:47:29.253Z] PASS __tests__/core/runner.test.ts (17.314 s)
[2020-12-02T11:47:29.512Z] PASS __tests__/dsl/journey.test.ts
[2020-12-02T11:47:29.512Z] 
[2020-12-02T11:47:29.512Z] Test Suites: 1 failed, 13 passed, 14 total
[2020-12-02T11:47:29.512Z] Tests:       43 passed, 43 total
[2020-12-02T11:47:29.512Z] Snapshots:   1 passed, 1 total
[2020-12-02T11:47:29.512Z] Time:        18.095 s
[2020-12-02T11:47:29.512Z] Ran all test suites.
[2020-12-02T11:47:30.662Z] Post stage
[2020-12-02T11:47:30.684Z] Recording test results
[2020-12-02T11:47:31.013Z] Stage "E2e Test" skipped due to earlier failure(s)
[2020-12-02T11:47:31.078Z] Stage "Publish Docker image" skipped due to earlier failure(s)
[2020-12-02T11:47:31.153Z] Stage "Matrix - STACK_VERSION = '8.0.0-synthetics'" skipped due to earlier failure(s)
[2020-12-02T11:47:31.159Z] Stage "Matrix - STACK_VERSION = '7.10.0-synthetics'" skipped due to earlier failure(s)
[2020-12-02T11:47:31.204Z] Stage "Matrix - STACK_VERSION = '8.0.0-synthetics'" skipped due to earlier failure(s)
[2020-12-02T11:47:31.211Z] Stage "Matrix - STACK_VERSION = '7.10.0-synthetics'" skipped due to earlier failure(s)
[2020-12-02T11:47:31.310Z] Stage "Matrix - STACK_VERSION = '8.0.0-synthetics'" skipped due to earlier failure(s)
[2020-12-02T11:47:31.317Z] Stage "Matrix - STACK_VERSION = '7.10.0-synthetics'" skipped due to earlier failure(s)
[2020-12-02T11:47:31.410Z] Stage "Matrix - STACK_VERSION = '8.0.0-synthetics'" skipped due to earlier failure(s)
[2020-12-02T11:47:31.417Z] Stage "Matrix - STACK_VERSION = '7.10.0-synthetics'" skipped due to earlier failure(s)
[2020-12-02T11:47:31.532Z] Failed in branch Matrix - STACK_VERSION = '8.0.0-synthetics'
[2020-12-02T11:47:31.540Z] Failed in branch Matrix - STACK_VERSION = '7.10.0-synthetics'
[2020-12-02T11:47:32.222Z] Running on Jenkins in /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-157
[2020-12-02T11:47:32.317Z] [INFO] getVaultSecret: Getting secrets
[2020-12-02T11:47:32.500Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-12-02T11:47:33.262Z] + chmod 755 generate-build-data.sh
[2020-12-02T11:47:33.262Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/elastic-synthetics/PR-157/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/elastic-synthetics/PR-157/runs/1 FAILURE 374351
[2020-12-02T11:47:33.813Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/elastic-synthetics/PR-157/runs/1/steps/?limit=10000 -o steps-info.json
[2020-12-02T11:47:34.363Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/elastic-synthetics/PR-157/runs/1/tests/?status=FAILED -o tests-errors.json

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

jahtalab reviewed (commented):
Thanks @vigneshshanmugam , A few points:
- AFAIK the `ts` field's origin is different depending on the platform it's running on, and it's not guaranteed to be monotonic.
- At the moment the user story around filmstrip is not clear (at least we decided to postpone it until we have a clear story)

Given these I suggest to keep the original issue open until we know how filmstrips is going to be used and therefore test properly according to that. 
Although we can merge this PR.

vigneshshanmugam commented on src/plugins/tracing.ts:
```diff
@@ -25,13 +25,26 @@
 
+/**
+ * Trace Event Format
+ * https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU
```
Not really, the trave events format hasn't changed and chrome devtool does not have any document related to trace events

jahtalab replied:
I think I saw a document for the Trace events but I can't find a link now.

vigneshshanmugam commented on src/plugins/tracing.ts:
```diff
@@ -25,13 +25,26 @@
   args?: {
     snapshot: string;
   };
+  // monotonic clock timestamp of the event
```
Yes based on the trace event library from Google dev https://github.com/samccone/chrome-trace-event/blob/d45bc8af3b5c53a3adfa2c5fc107b4fae054f579/lib/trace-event.ts#L22

vigneshshanmugam replied:
Forgot to link it, will do as part of the review

jahtalab replied:
But this only shows how that particular library has chosen to implement this (which I agree is the correct way), but it doesn't mean the event that we get from chrome is monotonic.

vigneshshanmugam replied:
It is indeed monotonic and depends on platform as you have mentioned, but we should be good here. I will add a comment about on this code and also the platforms i tested (mac and linux - our docker)

vigneshshanmugam commented:
Do you know the clock being different on platforms? Asking bcoz of that's the case we might have to change our strategy around gathering events from CDT.. Also we only support Linux atm, so this is a not a huge concern atleast IMO. 

We can also open a new issue  to enhance the filmstrip capturing if we want to do something different than the one planned now.

jahtalab reviewed (commented):
The platform difference is based on the chrome source code, (I will try to dig it up).
But I agree that linux should be our main objective, that's also why I'm suggesting to keep the issue open so that we don't invest more time on investigating this, at this point in time, but when we actually get around to implementing this feature.

vigneshshanmugam commented:
I have changed the issue description and will keep the issue open as per your suggestion, I have also digged up yesterday around how the timestamps and how chrome trace events would change in the future. Added the detailed comments here - https://github.com/elastic/synthetics/issues/85#issuecomment-737660628

Lets go ahead and merge this PR as the solution is good enough for the platforms we are aiming for now and works smoothly when tested on Mac and Linux.

vigneshshanmugam commented on src/helpers.ts:
```diff
@@ -52,14 +52,26 @@ export function generateTempPath() {
 
+/**
+ * Converts the trace events timestamp field from the
+ * format -  hrTime[0] * 1e6 + Math.round(hrTime[1] / 1000) to
```
Done.
